### PR TITLE
Dselans/catch err on startup

### PR DESCRIPTION
**Pull request by dselans**
@blinktag let's discuss this in the morning.

Most important part is that `New()` will now error on initial registration failure. This wasn't happening before unless the server was down - an auth error would not cause a failure because `.Register()` will only error on a transport error. This change checks the first recv - if it errors AND it's an initial registration - it will bail out and return an error.

I also exposed an `IgnoreStartupError` config flag that you can pass to "try initializing forever" (that is how it essentially worked before).

I also added some additional validation and comments/docs.

I also udpated the tests - they were actually returning a false negative before - as in, they were suceeding even though the server was not ever able to startup. This was masked by the fact that Register() never failed before so the test for New() never failed, even though it should have. To be able to perform a full gRPC server test, I needed to implement the `Register(...)` gRPC handler and have it send something to the client.

Aside from the above, I would like to discuss `SnitchURL` and `SnitchToken` once more.

EDIT: I'm going to leave some comments so that I don't forget why I did xyz 😄

### CODE REVIEW

File: register.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ func (s *Snitch) register(looper director.Looper) error {
 		ClientInfo: &protos.ClientInfo{
 			ClientType:     protos.ClientType(s.config.ClientType),
 			LibraryName:    "snitch-go-client",
-			LibraryVersion: "0.0.40",
+			LibraryVersion: "0.0.40", // TODO: Have CI bump this on release
```
Comment (dselans) [inline register.go:26]: Having to bump this manually sucks.

File: register.go [hunk 2]
```diff
@@ -36,12 +36,18 @@ func (s *Snitch) register(looper director.Looper) error {
...
+		err                error
+		quit               bool
+		initialRegister    = true
+		initialRegisterErr error
+	)
 
+	// This might not error even if the handler returns an err - need to attempt
+	// to perform a recv to verify.
 	srv, err := s.serverClient.Register(s.config.ShutdownCtx, req)
-	if err != nil && !strings.Contains(err.Error(), context.Canceled.Error()) {
```
Comment (dselans) [inline register.go:44]: Is there a reason why we shouldn't handle this as an err if context was cancelled?

File: register.go [hunk 3]
```diff
@@ -84,6 +93,17 @@ func (s *Snitch) register(looper director.Looper) error {
 		// Blocks until something is received
 		cmd, err := stream.Recv()
 		if err != nil {
+			// This is the first registration attempt and it has failed.
+			// Depending on IgnoreStartupError, we may need to stop the loop
+			// and tell the caller that we failed to complete registration.
+			if initialRegister && !s.config.IgnoreStartupError {
```
Comment (dselans) [inline register.go:99]: Without this, `New()` would only fail if `snitch-server` was unavailable. Auth, bad requests, etc. would not cause an err - need to perform a `Recv()` to get any other errors returned by the gRPC server.

File: server/client.go [hunk 4]
```diff
@@ -61,7 +61,7 @@ func New(snitchAddress, snitchToken string) (*Client, error) {
 
 	conn, err := grpc.DialContext(dialCtx, snitchAddress, opts...)
 	if err != nil {
-		return nil, errors.Wrap(err, "Could not dial GRPC server: %s")
+		return nil, errors.Wrap(err, "could not dial GRPC server: %s")
```
Comment (dselans) [inline server/client.go:64]: lowercase errors

File: snitch.go [hunk 5]
```diff
@@ -36,18 +36,18 @@ import (
...
 type OperationType int
 
+// ClientType is used to indicate if this library is being used by a shim or directly (as an SDK)
 type ClientType int
 
 const (
-	// OperationTypeConsumer tells Process to run the pipelines against the consume ruleset
-	OperationTypeConsumer OperationType = 1
+	// DefaultPipelineTimeoutDurationStr is the default timeout for a pipeline execution
+	DefaultPipelineTimeoutDurationStr = "100ms"
```
Comment (dselans) [inline snitch.go:47]: This and StepTimeout were hardcoded before.

File: snitch.go [hunk 6]
```diff
@@ -58,17 +58,26 @@ const (
...
-	// ClientTypeSDK & ClientTypeShim are referenced by shims and SDKs to specify
-	// which type of client they are upon registration with the snitch-server.
+	// ClientTypeSDK & ClientTypeShim are referenced by shims and SDKs to indicate
+	// in what context this SDK is being used.
 	ClientTypeSDK  ClientType = 1
 	ClientTypeShim ClientType = 2
+
+	// OperationTypeConsumer and OperationTypeProducer are used to indicate the
+	// type of operation the Process() call is performing.
+	OperationTypeConsumer OperationType = 1
```
Comment (dselans) [inline snitch.go:68]: The available/valid ints were not defined before so user had to specify `1` or `2`. I think?

File: snitch.go [hunk 7]
```diff
@@ -95,16 +104,55 @@ type Snitch struct {
...
-	ShutdownCtx     context.Context
-	Logger          logger.Logger
-	ClientType      ClientType
-	Audiences       []*Audience
+
+	// StepTimeout defines how long this library will allow a single step to run.
+	// Optional; default: 10ms
+	StepTimeout time.Duration
+
+	// IgnoreStartupError defines how to handle an error on initial startup via
```
Comment (dselans) [inline snitch.go:125]: Added explanations/comments/docs for everything you can tweak.

File: snitch.go [hunk 8]
```diff
@@ -278,6 +329,31 @@ func validateConfig(cfg *Config) error {
 		cfg.Logger = &logger.NoOpLogger{}
 	}
 
+	// Default to ClientTypeSDK
+	if cfg.ClientType != ClientTypeShim && cfg.ClientType != ClientTypeSDK {
+		cfg.ClientType = ClientTypeSDK
```
Comment (dselans) [inline snitch.go:334]: Is there any reason to not default the ClientType?

File: snitch_test.go [hunk 9]
```diff
@@ -93,6 +100,14 @@ func (i *InternalServer) GetAttachCommandsByService(ctx context.Context, req *pr
 	return &protos.GetAttachCommandsByServiceResponse{}, nil
 }
 
+func (i *InternalServer) Register(req *protos.RegisterRequest, srv protos.Internal_RegisterServer) error {
```
Comment (dselans) [inline snitch_test.go:103]: This had to be added for `New()` test to work.

File: register.go (hunk 1 above)
Comment (blinktag) [reply register.go:26]: you don't have to bump, CI does this already: https://github.com/streamdal/snitch-go-client/blob/main/.github/workflows/main-test.yml#L27-L45

File: register.go (hunk 2 above)
Comment (blinktag) [reply register.go:44]: This was previously logspamming on service shutdown. Behavior might have changed now that reconnect logic is different

File: snitch.go (hunk 8 above)
Comment (blinktag) [reply snitch.go:334]: Nope, should definitely have a default 👍

File: snitch.go (hunk 6 above)
Comment (blinktag) [reply snitch.go:68]: They were defined before here, https://github.com/streamdal/snitch-go-client/pull/26/files#diff-6b5a0e6cb3e11066b02351f5616b23cb0b08387a28f69cb7ea9552822cacaaa6L47

**Review by blinktag — APPROVED** [review]

File: register.go (hunk 1 above)
Comment (dselans) [reply register.go:26]: Oh shit. Thank you! That is excellent!
